druid-shell/src/platform/x11/xkb.rs

```diff
@@ -0,0 +1,456 @@
+        let key = self.get_logical_key(scancode);
+        // TODO this is lazy - really should use xkb i.e. augment the get_logical_key method.
+        let location = code_to_location(code);
+
+        let repeat = false;
+        // TODO not sure how to get this
+        let is_composing = false;
+
+        let mut mods = Modifiers::empty();
+        // Update xkb's state (e.g. return capitals if we've pressed shift)
+        unsafe {
+            xkb_state_update_key(
+                self.state,
+                scancode,
+                match state {
+                    KeyState::Down => XKB_KEY_DOWN,
+                    KeyState::Up => XKB_KEY_UP,
+                },
+            );
+            // compiler will unroll this loop
+            // FIXME(msrv): remove .iter().cloned() when msrv is >= 1.53
+            for (idx, mod_) in [
+                (self.mods.control, Modifiers::CONTROL),
+                (self.mods.shift, Modifiers::SHIFT),
+                (self.mods.super_, Modifiers::SUPER),
+                (self.mods.alt, Modifiers::ALT),
+                (self.mods.caps_lock, Modifiers::CAPS_LOCK),
+                (self.mods.num_lock, Modifiers::NUM_LOCK),
+            ]
+            .iter()
+            .cloned()
+            {
+                if xkb_state_mod_index_is_active(self.state, idx, XKB_STATE_MODS_EFFECTIVE) != 0 {
+                    mods |= mod_;
+                }
+            }
+        }
+        KeyEvent {
+            state,
+            key,
+            code,
+            location,
+            mods,
+            repeat,
+            is_composing,
+        }
+    }
+
+    fn get_logical_key(&self, scancode: u32) -> Key {
+        let mut key = map_key(self.key_get_one_sym(scancode));
+        if matches!(key, Key::Unidentified) {
+            if let Some(s) = self.key_get_utf8(scancode) {
+                key = Key::Character(s);
+            }
+        }
+        key
+    }
+
+    fn key_get_one_sym(&self, scancode: u32) -> u32 {
+        unsafe { xkb_state_key_get_one_sym(self.state, scancode) }
+    }
+
+    /// Get the string representation of a key.
+    // TODO `keyboard_types` forces us to return a String, but it would be nicer if we could stay
+    // on the stack, especially since we expect most results to be pretty small.
+    fn key_get_utf8(&self, scancode: u32) -> Option<String> {
+        unsafe {
+            // First get the size we will need
+            let len = xkb_state_key_get_utf8(self.state, scancode, ptr::null_mut(), 0);
+            if len == 0 {
+                return None;
+            }
+            // add 1 because we will get a null-terminated string.
+            let len = usize::try_from(len).unwrap() + 1;
+            let mut buf: Vec<u8> = Vec::new();
+            buf.resize(len, 0);
+            xkb_state_key_get_utf8(self.state, scancode, buf.as_mut_ptr() as *mut c_char, len);
+            debug_assert!(buf[buf.len() - 1] == 0);
```

Comment:
Maybe `assert!` instead of `debug_assert!` (because it's cheap and a possible source of memory errors)